text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,1062 @@
+
+```rust
+impl<'b,U> Iterator for Iter<'b,U> {
+    type Item = /* <TYPE> */;
+    ...
+}
+```
+
+So when we normalize `<PROJ>`, we obtain the result by applying some
+substitution `Θ` to `<TYPE>`. This substitution is a mapping from the
+lifetime/type parameters on the impl to some specific values, such
+that `<PROJ> == Θ <Iter<'b,U> as Iterator>::Item`. In this case, that
+means `Θ` would be `['b => 'a, U => T]` (and of course `<TYPE>` would
+be `&'b U`, but we're not supposed to rely on that).
+
+The key idea for the `OutlivesProjectionComponents` is that the only
+way that `<TYPE>` can *fail* to outlive `'x` is if either:
+
+- it names some lifetime parameter `'p` where `'p: 'x` does not hold; or,
+- it names some type parameter `X` where `X: 'x` does not hold.
+
+Now, the only way that `<TYPE>` can refer to a parameter `P` is if it
+is brought in by the substitution `Θ`. So, if we can just show that
+all the types/lifetimes that in the range of `Θ` outlive `'x`, then we
+know that `Θ <TYPE>` outlives `'x`.
+
+Put yet another way: imagine that you have an impl with *no
+parameters*, like:
+
+```rust
+impl Iterator for Foo {
+    type Item = /* <TYPE> */;
+}
+```
+
+Clearly, whatever `<TYPE>` is, it can only refer to the lifetime
+`'static`.  So `<Foo as Iterator>::Item: 'static` holds. We know this
+is true without ever knowing what `<TYPE>` is -- we just need to see
+that the trait reference `<Foo as Iterator>` doesn't have any
+lifetimes or type parameters in it, and hence the impl cannot refer to
+any lifetime or type parameters.
+
+#### Implementation complications
+
+One complication for the implementation is that there are so many
+potential outlives rules for projections. In particular, the rule that
+says `<P0 as Trait<P1..Pn>>>: 'a` holds if `Pi: 'a` is not an "if and
+only if" rule. So, for example, if we know that `T: 'a` and `'b: 'a`,
+then we know that `<T as Trait<'b>>:: Item: 'a` (for any trait and
+item), but not vice versa. This complicates inference significantly,
+since if variables are involved, we do not know whether to create
+edges between the variables or not (put another way, the simple
+dataflow model we are currently using doesn't truly suffice for these
+rules).
+
+This complication is unfortunate, but to a large extent already exists
+with where-clauses and trait matching (see e.g. [#21974]). (Moreover,
+it seems to be inherent to the concept of assocated types, since they
+take several inputs (the parameters to the trait) which may or may not
+be related to the actual type definition in question.)
+
+For the time being, the current implementation takes a pragmatic
+approach based on heuristics. It tries to avoid adding edges to the
+region graph in various common scenarios, and in the end falls back to
+enforcing conditions that may be stricter than necessary, but which
+certainly suffice. We have not yet encountered an example in practice
+where the current implementation rules do not suffice.
```

Comment:
@RalfJung 

> ...but doesn't know how the compiler works internally

I'll try to tweak the language somewhat to make it more accessible.